src/nodes/branch.rs

```diff
@@ -1,76 +1,191 @@
 use super::{super::TrieMask, rlp_node, CHILD_INDEX_RANGE};
-use alloy_primitives::B256;
-use alloy_rlp::{BufMut, EMPTY_STRING_CODE};
+use alloy_primitives::{hex, B256};
+use alloy_rlp::{length_of_length, Buf, BufMut, Decodable, Encodable, Header, EMPTY_STRING_CODE};
+use core::fmt;
 
 #[allow(unused_imports)]
-use alloc::{collections::BTreeMap, vec::Vec};
-
-/// A Branch node is only a pointer to the stack of nodes and is used to
-/// create the RLP encoding of the node using masks which filter from
-/// the stack of nodes.
-#[derive(Clone, Debug)]
-pub struct BranchNode<'a> {
-    /// Rlp encoded children
+use alloc::vec::Vec;
+
+/// A branch node in an Merkle Patricia Trie is a 17-element array consisting of 16 slots that
+/// correspond to each hexadecimal character and an additional slot for a value. We do exclude
+/// the node value since all paths have a fixed size.
+#[derive(PartialEq, Eq, Default)]
+pub struct BranchNode {
+    /// The collection of RLP encoded children.
+    pub stack: Vec<Vec<u8>>,
+    /// The bitmask indicating the presence of children at the respective nibble positions
+    pub state_mask: TrieMask,
+}
+
+impl fmt::Debug for BranchNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("BranchNode")
+            .field("stack", &self.stack.iter().map(hex::encode).collect::<Vec<_>>())
+            .field("state_mask", &self.state_mask)
+            .field("first_child_index", &self.as_ref().first_child_index())
+            .finish()
+    }
+}
+
+impl Encodable for BranchNode {
+    fn encode(&self, out: &mut dyn BufMut) {
+        self.as_ref().encode(out)
+    }
+
+    fn length(&self) -> usize {
+        self.as_ref().length()
+    }
+}
+
+impl Decodable for BranchNode {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let mut bytes = Header::decode_bytes(buf, true)?;
+
+        let mut stack = Vec::new();
+        let mut state_mask = TrieMask::default();
+        for index in CHILD_INDEX_RANGE {
+            // The buffer must contain empty string code for value.
+            if bytes.len() <= 1 {
+                return Err(alloy_rlp::Error::InputTooShort);
+            }
+
+            if bytes[0] == EMPTY_STRING_CODE {
```

Comment:
`Header::decode_bytes` should work here

Reply:
doesn't seem to :) my guess is that it's not really a header and first byte of a child can take on any value, but not debugging now